cmd/terramate/cli/cli.go

```diff
@@ -436,9 +462,13 @@ func (c *cli) printRunOrder() error {
 		stacks[i] = e.Stack
 	}
 
-	order, err := terramate.RunOrder(c.root(), stacks, c.parsedArgs.Changed)
+	order, reason, err := terramate.RunOrder(c.root(), stacks, c.parsedArgs.Changed)
 	if err != nil {
-		c.logerr("error: %v", err)
+		if errors.Is(err, dag.ErrCycleDetected) {
+			c.logerr("error: %v: reason: %s", err, reason)
+		} else {
+			c.logerr("error: %v", err)
+		}
```

Comment:
I thought about this but having some errors handled differently than others is also not nice. The reason, in this case, is not much different than the context we wrap with sentinel errors but I kept this separate in the API for easy testing... It'd also work if I wrap the reason like `errutil.Chain(ErrCycleDetected, reason)` and unwrap in tests to check but it seems messy and error-prone...

So yeah, I can add this type but if so users will need to know which methods can return a cycle error to have this logic, otherwise they will need to add it to each one even the ones that don't trigger this...

Reply:
Regarding "I can add this type but if so users will need to know which methods can return a cycle error to have this logic"

Well, now users need to know which methods return the sentinel terramate.ErrCycleDetected + they need to know the relationship between the ErrCycleDetected + the reason return value, no ? If not then the code could be just:

```go
if err != nil {
               if reason != "" {
			c.logerr("error: %v: reason: %s", err, reason)
		} else {
			c.logerr("error: %v", err)
		}
}
```

Reply:
```
func (d *DAG) Validate() (reason string, err error)

func RunOrder(root string, stacks []stack.S, changed bool) (stacks []stack.S, reason string, err error)
```

The `reason` is explicit in the APIs, so user only has to check for ErrCycleDetected in the same way it checks for other sentinel errors, and the reason comes separated when it's needed only.

By having the `ErrCycleDetected.Reason` in the error type, this is hidden behind the `error` interface.

Reply:
hmm so why the current code does check for the sentinel ?

```go
if errors.Is(err, dag.ErrCycleDetected) 
```

Shouldn't it just check the reason since it is explicit/decoupled ?

Reply:
I'm saying this because if reason should always exist only when it is an ErrCycleDetected then there is an advantage on making this explicit. The current API doesn't , nothing makes it clear that reason is only non-empty if the error is a specific sentinel (cycle err), could use docs but would prefer code itself making the relationship clear by adding an error type, the reason is for cycle detection problems only, so ErrCycleDetected.Reason is more clear/explicit IMHO. 

If the reason is not related only to cycle detection, then just checking the reason value itself (and not checking the sentinel) should be enough.

Reply:
We talked outside the PR, we are going to revisit the error handling topic later, lots to learn on this